File: 221/algoritms-master/lab3/test/task4.cpp
#include <bits/stdc++.h>
using namespace std;

void add_edge(vector<int> adj[], int src, int dest)
{
    adj[src].push_back(dest);
    adj[dest].push_back(src);
}

bool BFS(vector<int> adj[], int src, int dest, int v,
         int pred[], int dist[])
{
 //edike first e sob node e false
    list<int> queue;
    bool visited[v];
    for (int i = 0; i < v; i++)
    {
        visited[i] = false;
        dist[i] = INT_MAX;
        pred[i] = -1;
    }
    visited[src] = true;
    dist[src] = 0;
    queue.push_back(src);

    // standard BFS algorithm
    while (!queue.empty())
    {
        int u = queue.front();
        queue.pop_front();
        for (int i = 0; i < adj[u].size(); i++)
        {
            if (visited[adj[u][i]] == false)
            {
                visited[adj[u][i]] = true;
                dist[adj[u][i]] = dist[u] + 1;
                pred[adj[u][i]] = u;
                queue.push_back(adj[u][i]);

                // We stop BFS when we find
                // destination.
                if (adj[u][i] == dest)
                    return true;
            }
        }
    }

    return false;
}

void printShortestDistance(vector<int> adj[], int s,
                           int dest, int v)
{

    int pred[v], dist[v];

    if (BFS(adj, s, dest, v, pred, dist) == false)
    {
        cout << "Given source and destination"
             << " are not connected";
        return;
    }
    vector<int> path;
    int crawl = dest;
    path.push_back(crawl);
    while (pred[crawl] != -1)
    {
        path.push_back(pred[crawl]);
        crawl = pred[crawl];
    }
    cout << "Minimum number of bridges are : "
         << dist[dest];
}

int main()
{
    int T;
    cin >> T;
    for (int i = 0; i < T; i++)
    {
        int ver, edg;
        cin >> ver >> edg;

        vector<int> adj[ver];
        // vector<vector<int>>adj(ver, vector<int> (0, 0));
        for (int j = 0; j < edg; j++)
        {
            int a, b;
            cin >> a >> b;
            add_edge(adj, a - 1, b - 1);
        }
        printShortestDistance(adj, 0, ver - 1, ver);
        // here i used 0 and ver-1 is because i coded bst and other functions taking 0 to n-1 search
    }

    return 0;
}

File: 221/algoritms-master/Lab2/lab333/task4.py
# with open("input.txt") as f:
#     lines = f.readlines()
#
#     f.close()
#     vertices = int(lines[0])
#     adjacent_list = [[] for i in range(vertices)]
#     for i in range(2, len(lines)):
#         first, second = lines[i].split()
#         first, second = int(first), int(second)
#
#         adjacent_list[first - 1].append(second)
#
#     print(adjacent_list)
#
#
# with open("input.txt") as file:
#     vertices = int(file.readline())
#     edges = int(file.readline())
#     graph = {}
#     for i in range(edges):
#         source, destination = map(int, file.readline().split())
#         if source in graph:
#             graph[source].append(destination)
#         else:
#             graph[source] = [destination]
#     visited = set()
#     print(graph)
#
def add_edge(adj, src, dest):
    adj[src].append(dest)
    adj[dest].append(src)

def BFS(adj, src, dest, v, pred, dist):
    queue = []
    visited = [False for i in range(v)]
    for i in range(v):
        dist[i] = 1000000
        pred[i] = -1
        visited[src] = True
        dist[src] = 0; queue.append(src) #--------------------------------------put next code(while loop) from this indentation
        while (len(queue) != 0):
            u = queue[0]; queue.pop(0)
            for i in range(len(adj[u])):
                if (visited[adj[u][i]] == False):
                    visited[adj[u][i]] = True
                    dist[adj[u][i]] = dist[u] + 1
                    pred[adj[u][i]] = u
                    queue.append(adj[u][i])
                    if (adj[u][i] == dest):
                        return True
                    return False
def printShortestDistance(adj, s, dest, v):
    pred=[0 for i in range(v)]
    dist=[0 for i in range(v)]
    if (BFS(adj, s, dest, v, pred, dist) == False):
        print("Given source and destination are not connected")
        path = []
        crawl = dest
        crawl = dest
        path.append(crawl) #----------------------------put next code(while loop of pred[craw] one) from this indentation
        while (pred[crawl] != -1):
            path.append(pred[crawl])
            crawl = pred[crawl]
            print("Shortest path length is : " + str(dist[dest]))

if __name__=='__main__':
     T=int(input())
     while(T>0):
         ver=int(input())
         edg=int(input())
         adj = [[] for i in range(ver)]
         for j in range(edg):
             a=int(input())
             b=int(input())
             add_edge(adj, a-1, b-1)
             printShortestDistance(adj, 0, ver-1, ver)